controller/metrics/clustersecretupdater.go

```diff
@@ -0,0 +1,49 @@
+package metrics
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/argoproj/argo-cd/util/db"
+)
+
+type clusterSecretUpdater struct {
+	infoSource HasClustersInfo
+	lock       sync.Mutex
+	db         db.ArgoDB
+}
+
+func (c clusterSecretUpdater) Run(ctx context.Context) {
+	tick := time.Tick(metricsCollectionInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			break
+		case <-tick:
+			for _, info := range c.infoSource.GetClustersInfo() {
+				cluster, err := c.db.GetCluster(context.Background(), info.Server)
+				if err != nil {
+					continue
+				}
+				toUpdate := false
+				if info.K8SVersion != "" && info.K8SVersion != cluster.ServerVersion {
+					cluster.ServerVersion = info.K8SVersion
+					toUpdate = true
+				}
+				if info.LastCacheSyncTime != nil {
```

Comment:
We need to let the user see the real synchronization error. The PR https://github.com/argoproj/gitops-engine/pull/32 exposes synchronization error in ClusterInfo structure. 
Please use it to update cluster connection state:

```golang
if info. LastCacheSyncTime != nil {
  if info.SyncError == nil {
    // use ConnectionStatusSuccessful
  } else {
    // use ConnectionStatusFailed and info.SyncError.Error() as message
  }
} else {
  // use ConnectionStatusUnknown
}

```

Please save actual synchronization error `info.SyncError.Error` in the annotation so user could see it.